File: magic_8_ball/magic_8_ball.js
// Magic 8 Ball: A silly script to "answer" questions
// Copyright 2015  David Ulrich
// 
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
// 
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
// 
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.

process.stdin.resume();
process.stdin.setEncoding('utf8');

var answers = require("./answers");

var exit_codes = require("./exit_codes");

function do_magic(q) {
	var sum = 0;
	
	// who, what, where, why, when, how, how many|much
	var answer_type = "probability";
	if (q.indexOf("what") == 0 || q.indexOf("wat") == 0) {
		answer_type = "value";
	}
	else if (q.indexOf("when") == 0) {
		answer_type = "temporal";
	}
	else if (q.indexOf("why") == 0) {
		answer_type = "reason";
	}

	for(var i=0;i<q.length;i++) {
		sum += q.charCodeAt(i);
	}
	
	var answer_list = answers[answer_type];
	return answer_list[sum % answer_list.length];
}

var prompt = "What would you like to know? ";

process.stdout.write(prompt);
process.stdin.on("data", function (question) {
	question = question.toLowerCase().replace(/\n/g,"");
	
	if (exit_codes[question]) {
		process.exit();
	}
	
	console.log(do_magic(question));
	process.stdout.write(prompt);
});

File: magic_8_ball/magic_8_ball.cs
using System;
using System.Collections.Generic;
using Newtonsoft.Json;

namespace Magic.Eight.Ball {
	public class AnswerLists {
		public List<string> ProbabilityList;
		public List<string> ReasonList;
		public List<string> TemporalList;
		public List<string> ValueList;
	}
	
	public class magic_8_ball {
		private static AnswerLists answers;
		private static List<string> exit_codes;
		
		private static string do_magic(string q) {
			byte[] bytes;
			int sum = 0;
			List<string> answer_list;
			
			answer_list = answers.ProbabilityList;
			if (q.IndexOf("what") == 0 || q.IndexOf("wat") == 0) {
				answer_list = answers.ValueList;
			}
			else if (q.IndexOf("when") == 0) {
				answer_list = answers.TemporalList;
			}
			else if (q.IndexOf("why") == 0) {
				answer_list = answers.ReasonList;
			}
			
			bytes = System.Text.Encoding.ASCII.GetBytes(q);
			for(int i=0;i < bytes.Length;i++) {
				sum += bytes[i];
			}
			
			return answer_list[sum % answer_list.Count];
		}
		
		private static void LoadAnswerLists(string path) {
			List<string> current_list = null;
			string current_token;
			string answer_list_data = System.IO.File.ReadAllText("answers.json");
			
			JsonTextReader reader = new JsonTextReader(new System.IO.StringReader(answer_list_data));
			while (reader.Read()) {
				if (reader.Value == null) continue;
				
				if (reader.TokenType == JsonToken.PropertyName) {
					current_token = reader.Value.ToString();
					
					if (current_token == "probability") current_list = answers.ProbabilityList;
					else if (current_token == "reason") current_list = answers.ReasonList;
					else if (current_token == "temporal") current_list = answers.TemporalList;
					else if (current_token == "value") current_list = answers.ValueList;
				}
				else if (reader.TokenType == JsonToken.String && current_list != null) {
					current_list.Add(reader.Value.ToString());
				}
			}
		}
		
		private static List<string> LoadExitCodes(string path) {
			List<string> exit_codes = new List<string>();
			
			string exit_codes_data = System.IO.File.ReadAllText("exit_codes.json");
			
			JsonTextReader reader = new JsonTextReader(new System.IO.StringReader(exit_codes_data));
			while (reader.Read()) {
				if (reader.TokenType == JsonToken.PropertyName && reader.Value != null) {
					exit_codes.Add(reader.Value.ToString());
				}
			}
			
			return exit_codes;
		}
		
		public static void Main(string[] args) {
			answers = new AnswerLists();
			answers.ProbabilityList = new List<string>();
			answers.ReasonList = new List<string>();
			answers.TemporalList = new List<string>();
			answers.ValueList = new List<string>();
			
			LoadAnswerLists("answers.json");
			
			exit_codes = LoadExitCodes("exit_codes.json");
			
			string question;
			while (true) {
				Console.WriteLine("What would you like to know?");
				question = Console.ReadLine().ToLower();
				
				if (exit_codes.IndexOf(question) > -1) {
					break;
				}
				else {
					Console.WriteLine(do_magic(question));
				}
			}
		}
	}
}
